File: Question 0140 - Word Break II/src/Solution.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution
{
    private void solve(int start, int end, String s, List<String> curr, List<String> ans, Set<String> words)
    {
        if(end == s.length()-1)
        {
            if(words.contains(s.substring(start, end+1)))
            {
                curr.add(s.substring(start, end+1));
                ans.add(String.join(" ", curr));
                curr.remove(curr.size()-1);
            }
            return;
        }

        //take
        if(words.contains(s.substring(start, end+1)))
        {
            curr.add((s.substring(start, end+1)));
            solve(end+1, end+1, s, curr, ans, words);
            curr.remove(curr.size()-1);
        }

        //skip
        solve(start, end+1, s, curr, ans, words);
    }
    public List<String> wordBreak(String s, List<String> wordDict)
    {
        Set<String> set = new HashSet<>(wordDict);
        List<String> ans = new ArrayList<>();
        List<String> curr = new ArrayList<>();
        solve(0, 0, s, curr, ans, set);
        return ans;
    }
}

File: Question 0140 - Word Break II/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
private:
    string joinString(vector<string> &arr, string delimiter)
    {
        string ans;
        for (int i = 0; i < arr.size(); i++)
        {
            ans += arr[i];
            if (i + 1 < arr.size())
                ans += delimiter;
        }

        return ans;
    }

    void solve(string s, vector<string> &curr, vector<string> &ans, set<string> &st, int start)
    {
        if (start == s.size())
        {
            ans.push_back(joinString(curr, " "));
            return;
        }

        for (int i = start; i < s.size(); i++)
        {
            string word = s.substr(start, i - start + 1);

            if (st.count(word))
            {
                curr.push_back(word);
                solve(s, curr, ans, st, i + 1);
                curr.pop_back();
            }
        }
    }

public:
    vector<string> wordBreak(string s, vector<string> &wordDict)
    {
        set<string> st(wordDict.begin(), wordDict.end());
        vector<string> curr;
        vector<string> ans;
        solve(s, curr, ans, st, 0);
        return ans;
    }
};